Move test gateway client to clients package

joshklop opened the pull request:
Allows us to remove the testsource package.

## Description

## Types of changes

- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)

## Testing

**Requires testing**: Yes

**Did you write tests??**: existing tests are sufficient

## Documentation

**If this requires a documentation update, did you add one?** unnecessary

joshklop commented:
It might make sense to create a `clients/gateway` package (or `clients/feeder` to prevent clashing with `starknetdata/gateway`) since the `testdata` won't apply to the p2p client. I'll hold off on that for now and if anyone has a strong opinion please let me know in the reviews :)

IronGauntlets commented:
>It might make sense to create a clients/gateway package (or clients/feeder to prevent clashing with starknetdata/gateway) since the testdata won't apply to the p2p client. I'll hold off on that for now and if anyone has a strong opinion please let me know in the reviews :)

One thing to note is that `feeder gateway` and `gateway` are 2 different endpoints. `Feeder gateway` provide starknet information while `gateway` accepts new transactions. We should resolve these naming issues first.

Regarding packages, I would like to see something similar to the following:

```go
clients
  |-> feeder
  |-> gateway
starknetdata
  |-> feeder
  |-> p2p
  |-> starknetdata.go
```

IronGauntlets commented on clients/gateway.go:
```diff
@@ -60,6 +64,62 @@ func NopBackoff(d time.Duration) time.Duration {
 }
 
+// NewTestGatewayClient returns a client and a function to close a test server.
+func NewTestGatewayClient(network utils.Network) (*GatewayClient, func()) {
```
Remove stutter from `NewTestGatewayClient` and `NewGatewayClient` by renaming them to `NewTestGateway` and `NewGateway`, respectively.

joshklop commented:
That makes sense @IronGauntlets. In that case, would the package structure look like the following? 
```
clients
  |-> feeder
  |-> gateway
  |-> p2p
starknetdata
  |-> feeder
  |-> gateway
  |-> p2p
  |-> starknetdata.go
```
Since we will usually be importing `clients/...` and `starknetdata/...` in the same places, it might be helpful to make the package names slightly different, like `adaptfeeder`, `adaptp2p`, etc.. Kind of ugly, but more explicit than doing things like `feeder2`.

IronGauntlets commented:
> Since we will usually be importing clients/... and starknetdata/... in the same places, it might be helpful to make the package names slightly different, like adaptfeeder, adaptp2p, etc.. Kind of ugly, but more explicit than doing things like feeder2.

I would prefer to keep the above structure and just have an import alias to differentiate them when they are imported together. I really don't want to have the word `adapt` in the package name.

```go
import (
  ".../clients/feeder"
  adaptfeeder ".../starknetdata/feeder"
)
```

joshklop commented:
Finished and rebased, ready for another round of reviews. Pretty much all of the changes are simple renaming/refactoring, limited as much as possible to what is in the commit message. I changed most references from "gateway" to "feeder" so we can unambiguously refer to the "gateway" API in the future.

IronGauntlets commented on starknetdata/feeder/gateway.go:
```diff
@@ -11,10 +11,10 @@ import (
 
 type Gateway struct {
-	client *feeder.GatewayClient
+	client *feeder.FeederClient
```
`feeder.Feeder` is a sutter.

IronGauntlets reviewed (approved):
Nice work!

jelilat commented on starknetdata/feeder/feeder.go:
```diff
@@ -1,35 +1,28 @@
+// BlockByNumber gets the block for a given block number from the feeder,
 // then adapts it to the core.Block type.
-func (g *Gateway) BlockByNumber(ctx context.Context, blockNumber uint64) (*core.Block, error) {
+func (g *Feeder) BlockByNumber(ctx context.Context, blockNumber uint64) (*core.Block, error) {
```
Since we've changed the name to `Feeder`, we should probably also change this

```suggestion
func (f *Feeder) BlockByNumber(ctx context.Context, blockNumber uint64) (*core.Block, error) {
```

jelilat commented on starknetdata/feeder/feeder.go:
```diff
@@ -160,9 +153,9 @@ func adaptL2ToL1Message(response *clients.L2ToL1Message) *core.L2ToL1Message {
+// Transaction gets the transaction for a given transaction hash from the feeder,
 // then adapts it to the appropriate core.Transaction types.
-func (g *Gateway) Transaction(ctx context.Context, transactionHash *felt.Felt) (core.Transaction, error) {
+func (g *Feeder) Transaction(ctx context.Context, transactionHash *felt.Felt) (core.Transaction, error) {
```
Same as above
```suggestion
func (f *Feeder) Transaction(ctx context.Context, transactionHash *felt.Felt) (core.Transaction, error) {
```
